File: angularapp/src/app/services/trainer.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Trainer } from '../models/trainer.model';
import { PaginatedTrainers } from '../models/paginatedTrainers.model';
import { environment } from 'src/environments/environment';

@Injectable({
  providedIn: 'root'
})
export class TrainerService {

  // public apiUrl:string='http://localhost:3001/trainers';
  private baseUrl = environment.apiUrl;

  constructor(private http:HttpClient) { }

  getAllTrainers():Observable<any>
  {
    return this.http.get(`${this.baseUrl}/trainer`);
  }

  getTrainerById(trainerId:number):Observable<any>
  {
    return this.http.get(`${this.baseUrl}/trainer/${trainerId}`);
  }

  addTrainer(trainer:Trainer):Observable<any>
  {
    return this.http.post(`${this.baseUrl}/trainer`,trainer);
  }

  updateTrainer(trainerId:number,trainer:Trainer):Observable<any>
  {
    return this.http.put(`${this.baseUrl}/trainer/${trainerId}`,trainer);
  }

  deleteTrainer(trainerId:number):Observable<void>
  {
    return this.http.delete<void>(`${this.baseUrl}/trainer/${trainerId}`);
  }


  getTrainersByPages(page:number,size:number):Observable<PaginatedTrainers>
  {
    return this.http.get<any>(`${this.baseUrl}/trainers?page=${page}&size=${size}`);
  }
}

File: springapp/src/main/java/com/examly/springapp/service/TrainerService.java
package com.examly.springapp.service;

import java.util.List;
import java.util.Optional;

import org.springframework.data.domain.Page;
import org.springframework.stereotype.Service;

import com.examly.springapp.model.Trainer;
import com.examly.springapp.model.TrainerResponseDto;

@Service
public interface TrainerService {

    TrainerResponseDto addTrainer(Trainer trainer);

    Optional<TrainerResponseDto> getTrainerById(Long trainerId);

    List<TrainerResponseDto> getAllTrainers();

    TrainerResponseDto updateTrainer(Long trainerId,Trainer trainer);

    TrainerResponseDto deleteTrainer(Long trainerId);

    //pagination

    Page<TrainerResponseDto> getTrainersByPage(int page,int size);

    void addBulkTrainers(List<Trainer> trainers);



}
